lib/std/io/Reader.zig

```diff
@@ -324,20 +324,47 @@ pub fn isBytes(self: Self, slice: []const u8) anyerror!bool {
     return matches;
 }
 
+/// Read a struct from the stream.
+/// Only packed and extern structs are supported.
+/// Packed structs must have a `@bitSizeOf` that is a multiple of eight.
 pub fn readStruct(self: Self, comptime T: type) anyerror!T {
-    // Only extern and packed structs have defined in-memory layout.
-    comptime assert(@typeInfo(T).@"struct".layout != .auto);
-    var res: [1]T = undefined;
-    try self.readNoEof(mem.sliceAsBytes(res[0..]));
-    return res[0];
+    switch (@typeInfo(T).@"struct".layout) {
+        .auto => @compileError("readStruct only supports packed and extern structs."),
+        .@"extern" => {
+            var res: [1]T = undefined;
+            try self.readNoEof(mem.sliceAsBytes(res[0..]));
+            return res[0];
+        },
+        .@"packed" => {
+            var bytes: [@divExact(@bitSizeOf(T), 8)]u8 = undefined;
+            try self.readNoEof(&bytes);
+            return @bitCast(bytes);
+        },
+    }
 }
 
+/// Read a struct having the specified endianness into the host endianness representation.
+/// Only packed and extern structs are supported.
+/// Packed structs must have a `@bitSizeOf` that is a multiple of eight.
 pub fn readStructEndian(self: Self, comptime T: type, endian: std.builtin.Endian) anyerror!T {
-    var res = try self.readStruct(T);
-    if (native_endian != endian) {
-        mem.byteSwapAllFields(T, &res);
+    switch (@typeInfo(T).@"struct".layout) {
+        .auto => @compileError("readStructEndian only supports packed and extern structs."),
+        .@"extern" => {
+            var res = try self.readStruct(T);
+            if (native_endian != endian) {
+                mem.byteSwapAllFields(T, &res);
+            }
+            return res;
+        },
+        .@"packed" => {
+            var bytes: [@divExact(@bitSizeOf(T), 8)]u8 = undefined;
+            try self.readNoEof(&bytes);
+            if (native_endian != endian) {
+                mem.reverse(u8, &bytes);
```

Comment:
added more tests https://github.com/ziglang/zig/pull/21601/commits/d2d0b8f6916412aa5d2306576eb41e6afd147f6e

because packed struct fields are placed in the backing integer from the least to the most significant bits of the backing integer, only a single bytes reversal is required to convert from the little endian representation to the big endian representation, and vice-versa.

Reply:
> I think if such a struct was written on a machine with one endianness and read back on a different machine with the other endianness you'd get different results?

If big-endian machine uses `writeStruct` and little-endian machine uses `readStruct`, yes the two machine would be working with different data, and it would be a bug. 
For such implementations, they would both need to use `writeStructEndian` / `readStructEndian` and one of the machine would pay the performance cost of byte swapping. These serialization implementation details are not changed by this PR.

The existence of `writeStruct` is there for implementations that do not care about the host endianness (for example writing and reading back from memory, where the user knows the host endianness will not change and it is desirable to use the host endianness for maximum performance).